xml_converter/src/packaging_protobin.cpp

```diff
@@ -243,10 +244,47 @@ void write_protobuf_file(
         &state);
 }
 
+// Write protobuf per top level category
+void write_protobuf_file_per_category(
+    const string& marker_pack_root_directory,
+    const map<string, Category>* marker_categories,
+    const vector<Parseable*>* parsed_pois) {
+    std::map<string, std::vector<Parseable*>> category_to_pois;
+    ProtoWriterState state;
+    state.marker_pack_root_directory = marker_pack_root_directory;
+
+    for (size_t i = 0; i < parsed_pois->size(); i++) {
+        Parseable* parsed_poi = (*parsed_pois)[i];
+        if (parsed_poi->classname() == "POI") {
+            Icon* icon = dynamic_cast<Icon*>(parsed_poi);
+            // TODO(331): This is the wrong place to lowercase() the category and is hiding some crimes elsewhere
+            category_to_pois[lowercase(icon->category.category)].push_back(parsed_poi);
+        }
+        else if (parsed_poi->classname() == "Trail") {
+            Trail* trail = dynamic_cast<Trail*>(parsed_poi);
+            // TODO(331): This is the wrong place to lowercase() the category and is hiding some crimes elsewhere
+            category_to_pois[lowercase(trail->category.category)].push_back(parsed_poi);
+        }
+        else {
+            std::cout << "Unknown type" << std::endl;
+        }
+    }
+
+    for (map<string, Category>::const_iterator it = marker_categories->begin(); it != marker_categories->end(); it++) {
+        StringHierarchy category_filter;
+        category_filter.add_path({it->first}, true);
+        _write_protobuf_file(
+            join_file_paths(state.marker_pack_root_directory, it->first + ".guildpoint"),
+            category_filter,
+            marker_categories,
+            category_to_pois,
+            &state);
+    }
+}
+
```

Comment:
And then this function duplicates all of the logic of `write_protobuf_file` in order to access the earlier removed ability to pass in a `category_fiter`.

If the counter argument is the building of the `category_to_pois` dict for each file, aka: performance optimizations. I would necessarily have expected to see benchmarks somewhere in this PR

Reply:
This function actually duplicates `write_protobuf_file_per_map_id`. Since that is now 3 places, perhaps the creation of `category_to_pois` should be its own function. 

My other thought was that I am already iterating over the `marker_categories` map in order to get the name and would need a second iteration if I were to do it for the category filter first before this function is called.